### PR TITLE
Fix for  Masker family reloaded: NiftiSpheresMasker silently ignores voxels

### PR DESCRIPTION
**Pull request by himanshupathak21061998**
Fix for issue #1638 
@KamalakerDadi  I started with a fresh pull request and a test file with name test_nifti_spheres_maker.py is already there in test folder so I added a file with name test_niftispheresmasker.py

### CODE REVIEW

File: nilearn/input_data/nifti_spheres_masker.py [hunk 1]
```diff
@@ -36,8 +38,14 @@ def _apply_mask_and_get_affinity(seeds, niimg, radius, allow_overlap,
 
         X = masking._apply_mask_fmri(niimg, mask_img)
     else:
+        #Adding  warning message 
+        if np.isnan(np.sum(_safe_get_data(niimg))):
+            warnings.warn('The imgs you have fedded into fit_transform()' 
+                          'contains NaN values which are converted to zeroes ')
+            X = _safe_get_data(niimg,True).reshape([-1, niimg.shape[3]]).T
```
Comment (bthirion) [inline nilearn/input_data/nifti_spheres_masker.py]: pep8: space after comma. 
You should use a syntax checker.

File: nilearn/input_data/tests/test_niftispheresmasker.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import numpy as np
+import nibabel
+
+from nilearn.input_data import NiftiSpheresMasker
+#Preparing img with nans inside it 
+def generate_imgs():
+    affine = np.eye(4)
+
+    data_with_nans = np.random.random((10, 10, 10))
```
Comment (bthirion) [inline nilearn/input_data/tests/test_niftispheresmasker.py]: if you fill the values with nans, you shold not create it with random call: use e.g. zeros instead

File: nilearn/input_data/tests/test_niftispheresmasker.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    data_with_nans[indices] = data_without_nans[indices]
+    img = nibabel.Nifti1Image(data_with_nans, affine)
+    return img
+
+def test_is_nifti_spheres_masker_give_nans():
+    seed = [(7, 7, 7)]
+    img = generate_imgs()
+    # No interaction of seed with nans
+    masker = NiftiSpheresMasker(seeds=seed, radius=1.)
+    assert not np.isnan(np.sum(masker.fit_transform(img)))
```
Comment (bthirion) [inline nilearn/input_data/tests/test_niftispheresmasker.py]: I don't see why this one is useful.

**Comment by himanshupathak21061998** [comment]
Hey @bthirion I can't able to understand what is the cause of travis-ci failure because it it giving error in test_fetch_atlas_pauli_2017() method which is not related to changes done by me please help me out

**Comment by kchawla-pi** [comment]
@himanshupathak21061998 That was a spurious error. I reran the CI and its all green. Thanks for doing this!

File: nilearn/input_data/tests/test_niftispheresmasker.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import numpy as np
```
Comment (KamalakerDadi) [inline nilearn/input_data/tests/test_niftispheresmasker.py]: Please add these tests in the existing file nilearn/input_data/tests/test_nifti_spheres_masker.py.

File: nilearn/input_data/tests/test_niftispheresmasker.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import nibabel
+
+from nilearn.input_data import NiftiSpheresMasker
+#Preparing img with nans inside it 
+def generate_imgs():
```
Comment (KamalakerDadi) [inline nilearn/input_data/tests/test_niftispheresmasker.py]: generate_imgs -> generate_img

File: nilearn/input_data/tests/test_niftispheresmasker.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    affine = np.eye(4)
+
+    data_with_nans = np.zeros((10, 10, 10), dtype=np.float64)
```
Comment (KamalakerDadi) [inline nilearn/input_data/tests/test_niftispheresmasker.py]: Please reduce to float32.

File: nilearn/input_data/tests/test_niftispheresmasker.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    data_with_nans[indices] = data_without_nans[indices]
+    img = nibabel.Nifti1Image(data_with_nans, affine)
+    return img
+
+def test_is_nifti_spheres_masker_give_nans():
+    seed = [(7, 7, 7)]
+    img = generate_imgs()
+    # Interaction of seed with nans
+    masker = NiftiSpheresMasker(seeds=seed, radius=2.)
+    assert not np.isnan(np.sum(masker.fit_transform(img)))
```
Comment (KamalakerDadi) [inline nilearn/input_data/tests/test_niftispheresmasker.py]: Please use assert_false

File: nilearn/input_data/tests/test_niftispheresmasker.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    
+def test_nifti_spheres_masker_with_mask_img():
+    affine = np.eye(4)
+    mask = np.ones((9, 9, 9))
+    seed = [(7, 7, 7)] 
+    mask_img = nibabel.Nifti1Image(mask, affine)
+    img = generate_imgs()
+    # When mask image is provided, the seed interacts within the brain, so no nans.
+    masker = NiftiSpheresMasker(seeds=seed, radius=2., mask_img=mask_img)
+    assert not np.isnan(np.sum(masker.fit_transform(img)))
```
Comment (KamalakerDadi) [inline nilearn/input_data/tests/test_niftispheresmasker.py]: assert_false

File: nilearn/input_data/nifti_spheres_masker.py [hunk 9]
```diff
@@ -36,8 +38,14 @@ def _apply_mask_and_get_affinity(seeds, niimg, radius, allow_overlap,
 
         X = masking._apply_mask_fmri(niimg, mask_img)
     else:
+        #Adding  warning message 
+        if np.isnan(np.sum(_safe_get_data(niimg))):
+            warnings.warn('The imgs you have fedded into fit_transform()'
```
Comment (kchawla-pi) [inline nilearn/input_data/nifti_spheres_masker.py]: Change the warning message to "niimg fed into fit_transform() contains NaNs which will be converted to zeros"

Comment (KamalakerDadi) [reply nilearn/input_data/nifti_spheres_masker.py]: I would keep it ```imgs``` rather than ```niimg``` as per the doc string at fit_transform.

Comment (kchawla-pi) [reply nilearn/input_data/nifti_spheres_masker.py]: Okay I will update my comment.

**Comment by himanshupathak21061998** [comment]
Thanks for giving suggestions. It's nice to work with this org . I added test in test_nifti_shperes_masker.py

File: nilearn/input_data/nifti_spheres_masker.py [hunk 10]
```diff
@@ -38,9 +38,9 @@ def _apply_mask_and_get_affinity(seeds, niimg, radius, allow_overlap,
 
         X = masking._apply_mask_fmri(niimg, mask_img)
     else:
-        #Adding  warning message 
+        # Adding warning message
```
Comment (kchawla-pi) [inline nilearn/input_data/nifti_spheres_masker.py]: The code clearly tells us that a warning is being raised here and under what condition so the comment is unwarranted.

**Comment by himanshupathak21061998** [comment]
Is there anything more to be done on this issue?

**Comment by bthirion** [comment]
LGTM now. Thx.

**Comment by kchawla-pi** [comment]
@GaelVaroquaux @jeromedockes 
If you guys have any inputs, now is the time.
If not I'll merge in a few days.

**Comment by jeromedockes** [comment]
> @GaelVaroquaux @jeromedockes
> If you guys have any inputs, now is the time.
> If not I'll merge in a few days.

no comment on my side, if you think it's good to go merge it!

**Comment by kchawla-pi** [comment]
The PR merged before this is failing in CircleCI. I am reverting this and that one to forestall any potential (un)merge conflicts, and then will remerge this.

**Pull request merged**